### ISSUE TITLE
[@types/hapi__hapi(^19.0.3)] : The union type for method property of ServerRoute interface has string | string[] which seems redundant and causing type checking errors but not reporting

### ISSUE THREAD

**Issue by yedhink** [DefinitelyTyped/DefinitelyTyped#44806]
- [x] I tried using the `"@types/hapi__hapi": ^19.0.3` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] Definitions by: Rafael Souza Fijalkowski <https://github.com/rafaelsouzaf>
                 Justin Simms <https://github.com/jhsimms>
                 Simon Schick <https://github.com/SimonSchick>
                 Rodrigo Saboya <https://github.com/saboya>
                 Silas Rech <https://github.com/lenovouser>
  - Authors: @rafaelsouzaf , @jhsimms , @SimonSchick , @saboya , @lenovouser 

# TL;DR
What is the use of having ` string | string[]` in the union type of `method` in the [ServerRoute](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/hapi__hapi/index.d.ts#L3207) interface? The string literal `Util.HTTP_METHODS_PARTIAL` already defines all the possible `methods` like `GET, POST` etc. The problem occurs during type checking when defining a route. Maybe this is not an issue at all, but I will describe in detail the problem i have encountered.

# Context
Im very new with typescript. For learning purposes Im using hapijs(@hapi/hapi(^19.1.1) and @types/hapi__hapi), and was trying to setup the routes. The `server.route` method takes in a array of route configuration objects `<ServerRoute[]>`. In the `ServerRoute` interface, there is a property called `method` which has the following type

    export interface ServerRoute {
    /**
     * (required) the HTTP method. Typically one of 'GET', 'POST', 'PUT', 'PATCH', 'DELETE', or 'OPTIONS'. Any HTTP method is allowed, except for 'HEAD'. Use '*' to match against any HTTP method
     * (only when an exact match was not found, and any match with a specific method will be given a higher priority over a wildcard match). Can be assigned an array of methods which has the same
     * result as adding the same route with different methods manually.
     */
    method: Util.HTTP_METHODS_PARTIAL | Util.HTTP_METHODS_PARTIAL[] | string | string[];

where the `Util.HTTP_METHODS_PARTIAL` is:

    type HTTP_METHODS_PARTIAL =
    'GET'
    | 'POST'
    | 'PUT'
    | 'PATCH'
    | 'DELETE'
    | 'OPTIONS'
    | HTTP_METHODS_PARTIAL_LOWERCASE;

# What I have tried

This is how i create the routes:-

    server.route(<ServerRoute[]>[
        {
            "method": "GETzz", // here "GETzz" is not a valid UTIL.HTTP_METHODS_PARTIAL
            "path": "/",
            "handler": rootHandler().get,
        }
    ]);

I tried asserting by `"method": <Util.HTTP_METHODS_PARTIAL>"GETzz";`, which gives me auto-completions for that type inside quotes, but the value("GETzz") is still not checked to be part of `UTIL.HTTP_METHODS_PARTIAL` or not.

But take this simple example not part of hapijs:-

    type Methods = "valid1" | "valid2";
    const obj:{method:Methods} = {
            method: "inValidValue",
    }

The above would result in the error that i was hoping for, which's `Type '"inValidValue"' is not assignable to type 'Methods'.`

# What i think might be causing the problem
Take this example not part of hapi type definition:-
```js
 type Methods = "valid1" | "valid2";
 interface X {
        method : Methods | string,
  }
 const obj:X = {
        method: <Methods>"INVALID_VALUE", // -> no error detected
 }
```
Here too the assignment of an invalid value is not recognized, since union consists of types `string` and `Methods`, where `Methods` is a string literal itself, thus making any `string` value valid.

# Question
The question is how to enforce the property `method` to only be assigned values that exists in type `UTIL.HTTP_METHODS_PARTIAL`? Currently i can give any value like `GETzz` etc but still no error is thrown, until I run the server and a request is made to the corresponding `path`. The information from that error is unreadable and in no way helps in pointing out that, giving the value of `method` as `GETzz` is why the server failed. 
**So wouldn't removing the `string | string[]` from the `union type of `method` solve this problem? Are there any consequences by removing that? Or is there some trivial way to solve this in typescript?**

**Comment by SimonSchick** [comment]
You're not wrong here, this should probably be `| '*'` rather than ` | string | string[]`.
The type was probably made overly broad to be more forgiving (hapi normalises methods).
Feel free to submit a PR.

**Comment by yedhink** [comment]
@SimonSchick I can definitely do that. The change wouldn't be considered as major change(version wise) I believe. Also are there any corresponding docs which I have to update with the PR?

**Comment by SimonSchick** [comment]
It _is_ technically breaking but usually anything that improves type safety is not considered to be breaking.
I don't think there are any docs that need to be updated.

**Comment by yedhink** [comment]
@SimonSchick From my research i understood that non-standard http methods too exists. So my doubt is whether hapi supports adding non-standard http methods? From researching i understand that the inbuilt nodejs-http-parser only supports several standard http methods. So I have raised an issue in official hapijs repo. I think this PR should be done only after getting a valid response on whether validating all strings as a http method was a feature or bug. What do you suggest?

Relvant Links:

- HapiJs Offcial Source Code for Server Route method validation - https://github.com/hapijs/hapi/blob/master/lib/config.js#L306
- Nodejs http parser old :-  https://github.com/nodejs/node-v0.x-archive/blob/v0.10.29/deps/http_parser/http_parser.h#L87-119
- Nodejs http parser new:- https://github.com/nodejs/http-parser/blob/41d5c469be36fe8ca4e8aa55cc400918cee392fe/http_parser.c#L890

**Comment by yedhink** [comment]
@SimonSchick If in the case that hapi supports non-standard http methods too, then wouldn't converting the ServerRoute **interface to accept Generics** would solve the problem? or even something like this as mentioned by one among the community:-
```js
interface NonStandardHTTP {
  method: string
}

interface ValidHTTP extends NonStandardHTTP {
  method: "GET" | "POST"
}

const t: ValidHTTP = {
  method: "INVALID_METHOD"
}
```

**Comment by orta** [comment]
Hi thread, we're moving DefinitelyTyped to use [GitHub Discussions](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/53377) for conversations the `@types` modules in DefinitelyTyped.

To help with the transition, we're closing all issues which haven't had activity in the last 6 months, which includes this issue. If you think closing this issue is a mistake, please pop into the [TypeScript Community Discord](https://discord.gg/typescript) and mention the issue in the `definitely-typed` channel.
